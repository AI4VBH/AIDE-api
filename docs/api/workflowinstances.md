# Workflow Instances

## `PUT /workflwoinstances/:workflow_instance_id/executions/:execution_id/acknowledge`

Acknowledge task errors for a given execution of a workflow.

### Parameters

| Parameter Type | Label | Description |
|----------------|-------|-------------|
| URL | `:workflow_instance_id` | The workflow instance ID |
| URL | `:execution_id` | The task execution ID |

### Response

| Code | Description |
|------|-------------|
| `200` | Returns acknowledged `MonaiWorkflowInstance` |
| `404` | Workflow instance not found or task execution not found. |
| `500` | Server error. The response will be a [Problem details](https://datatracker.ietf.org/doc/html/rfc7807) object with server error details. |

### Example Request

```bash
curl --location --request PUT 'http://localhost:5000/workflwoinstances/3ef05fc1-dd5e-402a-b382-a557de1e383c/executions/04533a0b-d43d-47d4-8aac-1efaca39b778/acknowledge'
```

### Example Response

```json
{
  "acknowledged_workflow_errors": "",
  "ae_title": "MONAI",
  "bucket_id": "monaideploy",
  "id": "a67a7af7-068b-44b8-a81b-def7b3e5403b",
  "input_metadata": {},
  "payload_id": "d801bd68-381d-4ff3-95a8-85c46380e075",
  "start_time": "2022-09-22T12:45:13.102Z",
  "status": "Failed",
  "tasks": [
    {
      "acknowledged_task_errors": "2022-10-03T12:00:00.001Z",
      "execution_id": "3b9d94b9-4285-45d4-bea9-491fa62b8f91",
      "execution_stats": {},
      "input_artifacts": {
        "input-dicom": "d801bd68-381d-4ff3-95a8-85c46380e075/dcm",
      },
      "input_parameters": {},
      "next_timeout": "2022-09-22T12:45:13.102Z",
      "output_artifacts": {},
      "output_directory": "d801bd68-381d-4ff3-95a8-85c46380e075/workflows/a67a7af7-068b-44b8-a81b-def7b3e5403b/3b9d94b9-4285-45d4-bea9-491fa62b8f91",
      "previous_task_id": "",
      "reason": "Connection time-out",
      "result": {},
      "status": "failed",
      "task_id": "mean-pixel-calc",
      "task_plugin_arguments": {},
      "task_start_time": "2022-09-22T12:45:13.102Z",
      "task_type": "argo",
      "timeout_interval": 0,
      "workflow_instance_id": "a67a7af7-068b-44b8-a81b-def7b3e5403b"
    }
  ],
  "workflow_id": "0c262733-05a1-4811-9d9a-b5cadb5a475a"
}
```