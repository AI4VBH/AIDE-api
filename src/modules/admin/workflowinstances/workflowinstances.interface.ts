export interface MonaiWorkflowInstance {
  id: string;
  ae_title: string;
  workflow_id: string;
  workflow_name: string;
  start_time: string;
  payload_id: string;
  status: string;
  bucket_id: string;
  acknowledged_workflow_errors: string;
  tasks: MonaiWorkflowTask[];
  input_metadata: [{ [name: string]: string }];
}

export interface MonaiWorkflowTask {
  execution_id: string;
  workflow_instance_id: string;
  previous_task_id: string;
  task_id: string;
  task_type: string;
  task_start_time: string;
  status: string;
  reason: string;
  acknowledged_task_errors: string;
  input_artifacts: [{ [name: string]: string }];
  output_artifacts: [{ [name: string]: string }];
  output_directory: string;
  result: [{ [name: string]: string }];
  next_timeout: string;
  timeout_interval: number;
  execution_stats?: [{ [name: string]: string }];
}
