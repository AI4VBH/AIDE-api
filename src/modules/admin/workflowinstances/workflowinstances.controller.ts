import { Controller, Put, Inject, Param, UseFilters } from '@nestjs/common';
import ExternalServerExceptionFilter from 'shared/http/external-server-exception.filter';
import { WorkflowInstancesService } from './workflowinstances.service';

@Controller('workflowinstances')
@UseFilters(ExternalServerExceptionFilter)
export class WorkflowInstanceController {
  @Inject(WorkflowInstancesService)
  private readonly wfiService: WorkflowInstancesService;

  @Put(':workflow_instance_id/executions/:execution_id/acknowledge')
  acknowledgeTaskError(
    @Param('workflow_instance_id') workflow_instance_id,
    @Param('execution_id') execution_id,
  ) {
    return this.wfiService.acknowledgeTaskError(
      workflow_instance_id,
      execution_id,
    );
  }
}
