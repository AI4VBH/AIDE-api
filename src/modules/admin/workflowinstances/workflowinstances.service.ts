import { HttpService } from '@nestjs/axios';
import { Inject, Injectable } from '@nestjs/common';
import { lastValueFrom } from 'rxjs';
import { MonaiWorkflowInstance } from './workflowinstances.interface';

@Injectable()
export class WorkflowInstancesService {
  @Inject(HttpService)
  private readonly httpService: HttpService;
  private readonly CONTROLLER: string = 'workflowinstances';
  async acknowledgeTaskError(
    workflow_instance_id: string,
    execution_id: string,
  ): Promise<MonaiWorkflowInstance> {
    const response = await lastValueFrom(
      this.httpService.put<MonaiWorkflowInstance>(
        `${this.CONTROLLER}/${workflow_instance_id}/executions/${execution_id}/acknowledge`,
      ),
    );

    return response.data;
  }

  async getAcknowledgedTaskErrors(): Promise<MonaiWorkflowInstance[]> {
    const response = await lastValueFrom(
      this.httpService.get<MonaiWorkflowInstance[]>('workflowinstances/failed'),
    );

    return response.data;
  }
}
