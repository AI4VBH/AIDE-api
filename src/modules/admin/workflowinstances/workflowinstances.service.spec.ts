import { createMock, DeepMocked } from '@golevelup/ts-jest';
import { HttpService } from '@nestjs/axios';
import { Test, TestingModule } from '@nestjs/testing';
import { AxiosInstance } from 'axios';
import { makeObservableForTest } from 'test/utilities/test-make-observable';
import { WorkflowInstancesService } from './workflowinstances.service';
import { MonaiWorkflowInstance } from './workflowinstances.interface';

describe('PayloadsService', () => {
  let axios: DeepMocked<AxiosInstance>;
  let httpService: DeepMocked<HttpService>;
  let service: WorkflowInstancesService;

  beforeEach(async () => {
    axios = createMock<AxiosInstance>();
    httpService = createMock<HttpService>({
      axiosRef: axios,
    });

    const module: TestingModule = await Test.createTestingModule({
      providers: [
        WorkflowInstancesService,
        {
          provide: HttpService,
          useValue: httpService,
        },
      ],
    }).compile();

    service = module.get<WorkflowInstancesService>(WorkflowInstancesService);
  });

  it('should be defined', () => expect(service).toBeDefined());
  describe('acknowledgeTaskError', () => {
    it('returns the expected result', async () => {
      const expectedResult: MonaiWorkflowInstance = {
        id: 'c5e980ff-e278-441b-af64-8630bf57b7e3',
        ae_title: 'some-title',
        payload_id: '041293d0-ab97-4ea1-b967-42ec62f26608',
        workflow_id: 'fca60a4d-2932-4196-9206-cf123c82332b',
        start_time: '2022-09-22T12:45:13.102Z',
        acknowledged_workflow_errors: '2022-09-22T12:45:13.102Z',
        status: 'Succeeded',
        workflow_name: 'lillie',
        tasks: [
          {
            task_id: 'email-task',
            workflow_instance_id: 'c5e980ff-e278-441b-af64-8630bf57b7e3',
            execution_id: '981eca65-294a-434a-930a-b1828b54253a',
            previous_task_id: 'export-task',
            task_start_time: '2022-09-23T12:50:00.305Z',
            task_type: 'test',
            status: 'Failed',
            acknowledged_task_errors: '2022-09-23T12:50:00.305Z',
            reason: 'undefined',
            input_artifacts: [{ name: '', value: '' }],
            output_artifacts: [{ name: '', value: '' }],
            output_directory: 'outputdir',
            result: [{ name: '', value: '' }],
            next_timeout: '2',
            timeout_interval: 60,
          },
        ],
        bucket_id: 'bucket',
        input_metadata: [{ name: '', value: '' }],
      };

      axios.put.mockResolvedValue({ data: expectedResult });

      httpService.put.mockReturnValue(makeObservableForTest(axios.put));

      const response = await service.acknowledgeTaskError(
        'c5e980ff-e278-441b-af64-8630bf57b7e3',
        '981eca65-294a-434a-930a-b1828b54253a',
      );

      expect(response).toMatchSnapshot();
    });

    it('gets a value returns valid response', async () => {
      const expectedResult: MonaiWorkflowInstance = {
        id: 'c5e980ff-e278-441b-af64-8630bf57b7e3',
        ae_title: 'some-title',
        payload_id: '041293d0-ab97-4ea1-b967-42ec62f26608',
        workflow_id: 'fca60a4d-2932-4196-9206-cf123c82332b',
        start_time: '2022-09-22T12:45:13.102Z',
        acknowledged_workflow_errors: '2022-09-22T12:45:13.102Z',
        status: 'Succeeded',
        workflow_name: 'lillie',
        tasks: [
          {
            task_id: 'email-task',
            workflow_instance_id: 'c5e980ff-e278-441b-af64-8630bf57b7e3',
            execution_id: '981eca65-294a-434a-930a-b1828b54253a',
            previous_task_id: 'export-task',
            task_start_time: '2022-09-23T12:50:00.305Z',
            task_type: 'test',
            status: 'Failed',
            acknowledged_task_errors: '2022-09-23T12:50:00.305Z',
            reason: 'undefined',
            input_artifacts: [{ name: '', value: '' }],
            output_artifacts: [{ name: '', value: '' }],
            output_directory: 'outputdir',
            result: [{ name: '', value: '' }],
            next_timeout: '2',
            timeout_interval: 60,
          },
        ],
        bucket_id: 'bucket',
        input_metadata: [{ name: '', value: '' }],
      };
      axios.get.mockResolvedValue({
        status: 200,
        data: [expectedResult],
      });
      httpService.get.mockReturnValue(makeObservableForTest(axios.get));

      const dateTimeNow = new Date().toISOString().split('T')[0];

      const response = await service.getAcknowledgedTaskErrors(dateTimeNow);

      expect(response).toMatchSnapshot();
    });
  });
});
