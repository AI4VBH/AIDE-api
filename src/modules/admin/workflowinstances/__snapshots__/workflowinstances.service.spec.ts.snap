// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PayloadsService acknowledgeTaskError gets a value returns valid response 1`] = `
[
  {
    "acknowledged_workflow_errors": "2022-09-22T12:45:13.102Z",
    "ae_title": "some-title",
    "bucket_id": "bucket",
    "id": "c5e980ff-e278-441b-af64-8630bf57b7e3",
    "input_metadata": [
      {
        "name": "",
        "value": "",
      },
    ],
    "payload_id": "041293d0-ab97-4ea1-b967-42ec62f26608",
    "start_time": "2022-09-22T12:45:13.102Z",
    "status": "Succeeded",
    "tasks": [
      {
        "acknowledged_task_errors": "2022-09-23T12:50:00.305Z",
        "execution_id": "981eca65-294a-434a-930a-b1828b54253a",
        "input_artifacts": [
          {
            "name": "",
            "value": "",
          },
        ],
        "next_timeout": "2",
        "output_artifacts": [
          {
            "name": "",
            "value": "",
          },
        ],
        "output_directory": "outputdir",
        "previous_task_id": "export-task",
        "reason": "undefined",
        "result": [
          {
            "name": "",
            "value": "",
          },
        ],
        "status": "Failed",
        "task_id": "email-task",
        "task_start_time": "2022-09-23T12:50:00.305Z",
        "task_type": "test",
        "timeout_interval": 60,
        "workflow_instance_id": "c5e980ff-e278-441b-af64-8630bf57b7e3",
      },
    ],
    "workflow_id": "fca60a4d-2932-4196-9206-cf123c82332b",
    "workflow_name": "lillie",
  },
]
`;

exports[`PayloadsService acknowledgeTaskError returns the expected result 1`] = `
{
  "acknowledged_workflow_errors": "2022-09-22T12:45:13.102Z",
  "ae_title": "some-title",
  "bucket_id": "bucket",
  "id": "c5e980ff-e278-441b-af64-8630bf57b7e3",
  "input_metadata": [
    {
      "name": "",
      "value": "",
    },
  ],
  "payload_id": "041293d0-ab97-4ea1-b967-42ec62f26608",
  "start_time": "2022-09-22T12:45:13.102Z",
  "status": "Succeeded",
  "tasks": [
    {
      "acknowledged_task_errors": "2022-09-23T12:50:00.305Z",
      "execution_id": "981eca65-294a-434a-930a-b1828b54253a",
      "input_artifacts": [
        {
          "name": "",
          "value": "",
        },
      ],
      "next_timeout": "2",
      "output_artifacts": [
        {
          "name": "",
          "value": "",
        },
      ],
      "output_directory": "outputdir",
      "previous_task_id": "export-task",
      "reason": "undefined",
      "result": [
        {
          "name": "",
          "value": "",
        },
      ],
      "status": "Failed",
      "task_id": "email-task",
      "task_start_time": "2022-09-23T12:50:00.305Z",
      "task_type": "test",
      "timeout_interval": 60,
      "workflow_instance_id": "c5e980ff-e278-441b-af64-8630bf57b7e3",
    },
  ],
  "workflow_id": "fca60a4d-2932-4196-9206-cf123c82332b",
  "workflow_name": "lillie",
}
`;
